### DISCUSSION TITLE
Why is AddCommandLine(args) called twice in CreateDefaultBuilder(string[] args)？

### DISCUSSION THREAD

**Discussion by Surbowl** [dotnet/aspnetcore#26733] [Q&A] [answered]
I find that this method is called twice in **WebHost.cs** -> **CreateDefaultBuilder(string[] args)**
https://github.com/dotnet/aspnetcore/blob/450e047fddd796a3ab764f48662d80a072e94dd4/src/DefaultBuilder/src/WebHost.cs#L164-L167
and
https://github.com/dotnet/aspnetcore/blob/450e047fddd796a3ab764f48662d80a072e94dd4/src/DefaultBuilder/src/WebHost.cs#L187-L190
But I see the following code in **WebHostBuilder.cs** -> **BuildCommonServices(out AggregateException? hostingStartupErrors)**
https://github.com/dotnet/aspnetcore/blob/450e047fddd796a3ab764f48662d80a072e94dd4/src/Hosting/Hosting/src/WebHostBuilder.cs#L271-L273
The new **builder** seems to contain the value of  **_config**, and **_config** already has CommandLine config in it, so I don't quite understand why it's called twice.
I'm sorry my English isn't very good😳

**Comment by Tratcher** [comment, marked as answer]
There are two stages to config. In the first one a few sources are loaded so the host itself can be configured. This is used to read things like ASPNETCORE_ENVIRONMENT, ContentRoot, etc..
https://github.com/dotnet/aspnetcore/blob/b7b300143518f86d5e4ff6630688c7d578efda07/src/DefaultBuilder/src/WebHost.cs#L164-L167

Later the host builds a config for the application. It starts with a copy of the host config so the app can re-use those values, and layers everything else on top of that. This is where things like AppSettings.Development.json are added, using the environment and the content root from the host config.
https://github.com/dotnet/aspnetcore/blob/b7b300143518f86d5e4ff6630688c7d578efda07/src/Hosting/Hosting/src/WebHostBuilder.cs#L271-L275

**Reply by Surbowl** [reply]
Thank you for your answer! but I still don't quite understand what this code does
https://github.com/dotnet/aspnetcore/blob/450e047fddd796a3ab764f48662d80a072e94dd4/src/DefaultBuilder/src/WebHost.cs#L187-L190
the application config is a copy of the host config, and host config already contains args, so why add args again in configureDelegate😳

**Reply by Tratcher** [reply]
Ah, that one is for precedence. Sources added later override the earlier ones, so command line args were added again so they could override  things like appsettings.json.

**Reply by Surbowl** [reply]
Thank you Tratcher. I get it ❤
